starting by importing a set of libraries 

In [36]:
import numpy as np
import pandas as pd
import urllib
from urllib.request import urlopen
from bs4 import BeautifulSoup
import bs4 as bs
%matplotlib inline


here the Toronto Postal Code website to be scraped is being called with pandas into a dataframe

In [37]:
dfs = pd.read_html('https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M', header=0)

In [38]:
dfs

[    Postcode           Borough  \
 0        M1A      Not assigned   
 1        M2A      Not assigned   
 2        M3A        North York   
 3        M4A        North York   
 4        M5A  Downtown Toronto   
 5        M5A  Downtown Toronto   
 6        M6A        North York   
 7        M6A        North York   
 8        M7A      Queen's Park   
 9        M8A      Not assigned   
 10       M9A         Etobicoke   
 11       M1B       Scarborough   
 12       M1B       Scarborough   
 13       M2B      Not assigned   
 14       M3B        North York   
 15       M4B         East York   
 16       M4B         East York   
 17       M5B  Downtown Toronto   
 18       M5B  Downtown Toronto   
 19       M6B        North York   
 20       M7B      Not assigned   
 21       M8B      Not assigned   
 22       M9B         Etobicoke   
 23       M9B         Etobicoke   
 24       M9B         Etobicoke   
 25       M9B         Etobicoke   
 26       M9B         Etobicoke   
 27       M1C       

Looking at the content and structure of the data

In [39]:
dfs[0].columns

Index(['Postcode', 'Borough', 'Neighbourhood'], dtype='object')

In [40]:
dfs[0].head(12) 

Postcode           Borough     Neighbourhood
0       M1A      Not assigned      Not assigned
1       M2A      Not assigned      Not assigned
2       M3A        North York         Parkwoods
3       M4A        North York  Victoria Village
4       M5A  Downtown Toronto      Harbourfront
5       M5A  Downtown Toronto       Regent Park
6       M6A        North York  Lawrence Heights
7       M6A        North York    Lawrence Manor
8       M7A      Queen's Park      Not assigned
9       M8A      Not assigned      Not assigned
10      M9A         Etobicoke  Islington Avenue
11      M1B       Scarborough             Rouge

In [41]:
dfs[0].shape

(288, 3)

using the boolean to figure out where in the Borough column data "Not assigned" values occur

In [42]:
import re
booleans = []
for result in dfs[0].Borough:
    if not re.search('Not assigned', result):
        booleans.append(True)
    else:
        booleans.append(False)

print(booleans[0:5])
print(len(booleans))


[False, False, True, True, True]
288


In [43]:
len(booleans)

288

Filtering out "Not assigned" values from the dataframe and leaving behind the new dataframe having 211 rows from the total 288

In [44]:
Filtered = pd.Series(booleans)
New_df = dfs[0][Filtered]
New_df.shape


(211, 3)

More than one neighborhood with the same postcode are put together in one row under the paticular postcode

In [45]:
dat = New_df.groupby(['Postcode','Borough'])['Neighbourhood'].apply(',' .join).reset_index()

In [46]:
dat

Postcode           Borough  \
0        M1B       Scarborough   
1        M1C       Scarborough   
2        M1E       Scarborough   
3        M1G       Scarborough   
4        M1H       Scarborough   
5        M1J       Scarborough   
6        M1K       Scarborough   
7        M1L       Scarborough   
8        M1M       Scarborough   
9        M1N       Scarborough   
10       M1P       Scarborough   
11       M1R       Scarborough   
12       M1S       Scarborough   
13       M1T       Scarborough   
14       M1V       Scarborough   
15       M1W       Scarborough   
16       M1X       Scarborough   
17       M2H        North York   
18       M2J        North York   
19       M2K        North York   
20       M2L        North York   
21       M2M        North York   
22       M2N        North York   
23       M2P        North York   
24       M2R        North York   
25       M3A        North York   
26       M3B        North York   
27       M3C        North York   
28       M3H        North York   
29       M3J        North York   
..       ...               ...   
73       M6C              York   
74       M6E              York   
75       M6G  Downtown Toronto   
76       M6H      West Toronto   
77       M6J      West Toronto   
78       M6K      West Toronto   
79       M6L        North York   
80       M6M              York   
81       M6N              York   
82       M6P      West Toronto   
83       M6R      West Toronto   
84       M6S      West Toronto   
85       M7A      Queen's Park   
86       M7R       Mississauga   
87       M7Y      East Toronto   
88       M8V         Etobicoke   
89       M8W         Etobicoke   
90       M8X         Etobicoke   
91       M8Y         Etobicoke   
92       M8Z         Etobicoke   
93       M9A         Etobicoke   
94       M9B         Etobicoke   
95       M9C         Etobicoke   
96       M9L        North York   
97       M9M        North York   
98       M9N              York   
99       M9P         Etobicoke   
100      M9R         Etobicoke   
101      M9V         Etobicoke   
102      M9W         Etobicoke   

                                         Neighbourhood  
0                                        Rouge,Malvern  
1                 Highland Creek,Rouge Hill,Port Union  
2                      Guildwood,Morningside,West Hill  
3                                               Woburn  
4                                            Cedarbrae  
5                                  Scarborough Village  
6            East Birchmount Park,Ionview,Kennedy Park  
7                        Clairlea,Golden Mile,Oakridge  
8        Cliffcrest,Cliffside,Scarborough Village West  
9                           Birch Cliff,Cliffside West  
10   Dorset Park,Scarborough Town Centre,Wexford He...  
11                                    Maryvale,Wexford  
12                                           Agincourt  
13               Clarks Corners,Sullivan,Tam O'Shanter  
14   Agincourt North,L'Amoreaux East,Milliken,Steel...  
15                                     L'Amoreaux West  
16                                         Upper Rouge  
17                                   Hillcrest Village  
18                          Fairview,Henry Farm,Oriole  
19                                     Bayview Village  
20                             Silver Hills,York Mills  
21                              Newtonbrook,Willowdale  
22                                    Willowdale South  
23                                     York Mills West  
24                                     Willowdale West  
25                                           Parkwoods  
26                                     Don Mills North  
27                     Flemingdon Park,Don Mills South  
28       Bathurst Manor,Downsview North,Wilson Heights  
29                      Northwood Park,York University  
..                                                 ...  
73                                  Humewood-Cedarvale  
74            

Not assigned name for Neighbourhood are replaced with the corresponding Borough names

In [47]:
dat['Neighbourhood'] = np.where(dat['Neighbourhood'] == 'Not assigned', dat['Borough'], dat['Neighbourhood'])

In [48]:
dat

Postcode           Borough  \
0        M1B       Scarborough   
1        M1C       Scarborough   
2        M1E       Scarborough   
3        M1G       Scarborough   
4        M1H       Scarborough   
5        M1J       Scarborough   
6        M1K       Scarborough   
7        M1L       Scarborough   
8        M1M       Scarborough   
9        M1N       Scarborough   
10       M1P       Scarborough   
11       M1R       Scarborough   
12       M1S       Scarborough   
13       M1T       Scarborough   
14       M1V       Scarborough   
15       M1W       Scarborough   
16       M1X       Scarborough   
17       M2H        North York   
18       M2J        North York   
19       M2K        North York   
20       M2L        North York   
21       M2M        North York   
22       M2N        North York   
23       M2P        North York   
24       M2R        North York   
25       M3A        North York   
26       M3B        North York   
27       M3C        North York   
28       M3H        North York   
29       M3J        North York   
..       ...               ...   
73       M6C              York   
74       M6E              York   
75       M6G  Downtown Toronto   
76       M6H      West Toronto   
77       M6J      West Toronto   
78       M6K      West Toronto   
79       M6L        North York   
80       M6M              York   
81       M6N              York   
82       M6P      West Toronto   
83       M6R      West Toronto   
84       M6S      West Toronto   
85       M7A      Queen's Park   
86       M7R       Mississauga   
87       M7Y      East Toronto   
88       M8V         Etobicoke   
89       M8W         Etobicoke   
90       M8X         Etobicoke   
91       M8Y         Etobicoke   
92       M8Z         Etobicoke   
93       M9A         Etobicoke   
94       M9B         Etobicoke   
95       M9C         Etobicoke   
96       M9L        North York   
97       M9M        North York   
98       M9N              York   
99       M9P         Etobicoke   
100      M9R         Etobicoke   
101      M9V         Etobicoke   
102      M9W         Etobicoke   

                                         Neighbourhood  
0                                        Rouge,Malvern  
1                 Highland Creek,Rouge Hill,Port Union  
2                      Guildwood,Morningside,West Hill  
3                                               Woburn  
4                                            Cedarbrae  
5                                  Scarborough Village  
6            East Birchmount Park,Ionview,Kennedy Park  
7                        Clairlea,Golden Mile,Oakridge  
8        Cliffcrest,Cliffside,Scarborough Village West  
9                           Birch Cliff,Cliffside West  
10   Dorset Park,Scarborough Town Centre,Wexford He...  
11                                    Maryvale,Wexford  
12                                           Agincourt  
13               Clarks Corners,Sullivan,Tam O'Shanter  
14   Agincourt North,L'Amoreaux East,Milliken,Steel...  
15                                     L'Amoreaux West  
16                                         Upper Rouge  
17                                   Hillcrest Village  
18                          Fairview,Henry Farm,Oriole  
19                                     Bayview Village  
20                             Silver Hills,York Mills  
21                              Newtonbrook,Willowdale  
22                                    Willowdale South  
23                                     York Mills West  
24                                     Willowdale West  
25                                           Parkwoods  
26                                     Don Mills North  
27                     Flemingdon Park,Don Mills South  
28       Bathurst Manor,Downsview North,Wilson Heights  
29                      Northwood Park,York University  
..                                                 ...  
73                                  Humewood-Cedarvale  
74            

the remaining total number of rows after the above exercise is performed are given as below

In [49]:
dat.shape

(103, 3)

Importing csv file for the coordinates of the city of Toronto and previewing it

In [50]:

data = pd.read_csv("Geospatial_coordinates.csv") 

In [51]:
data.head(11)

Postcode   Latitude  Longitude
0       M1B  43.806686 -79.194353
1       M1C  43.784535 -79.160497
2       M1E  43.763573 -79.188711
3       M1G  43.770992 -79.216917
4       M1H  43.773136 -79.239476
5       M1J  43.744734 -79.239476
6       M1K  43.727929 -79.262029
7       M1L  43.711112 -79.284577
8       M1M  43.716316 -79.239476
9       M1N  43.692657 -79.264848
10      M1P  43.757410 -79.273304

Merging the coordinates together with the city names in Toronto

In [52]:
#pd.concat([dat,data], axis=1)
new_data = dat.merge(data, 'outer', 'Postcode')
new_data

Postcode           Borough  \
0        M1B       Scarborough   
1        M1C       Scarborough   
2        M1E       Scarborough   
3        M1G       Scarborough   
4        M1H       Scarborough   
5        M1J       Scarborough   
6        M1K       Scarborough   
7        M1L       Scarborough   
8        M1M       Scarborough   
9        M1N       Scarborough   
10       M1P       Scarborough   
11       M1R       Scarborough   
12       M1S       Scarborough   
13       M1T       Scarborough   
14       M1V       Scarborough   
15       M1W       Scarborough   
16       M1X       Scarborough   
17       M2H        North York   
18       M2J        North York   
19       M2K        North York   
20       M2L        North York   
21       M2M        North York   
22       M2N        North York   
23       M2P        North York   
24       M2R        North York   
25       M3A        North York   
26       M3B        North York   
27       M3C        North York   
28       M3H        North York   
29       M3J        North York   
..       ...               ...   
73       M6C              York   
74       M6E              York   
75       M6G  Downtown Toronto   
76       M6H      West Toronto   
77       M6J      West Toronto   
78       M6K      West Toronto   
79       M6L        North York   
80       M6M              York   
81       M6N              York   
82       M6P      West Toronto   
83       M6R      West Toronto   
84       M6S      West Toronto   
85       M7A      Queen's Park   
86       M7R       Mississauga   
87       M7Y      East Toronto   
88       M8V         Etobicoke   
89       M8W         Etobicoke   
90       M8X         Etobicoke   
91       M8Y         Etobicoke   
92       M8Z         Etobicoke   
93       M9A         Etobicoke   
94       M9B         Etobicoke   
95       M9C         Etobicoke   
96       M9L        North York   
97       M9M        North York   
98       M9N              York   
99       M9P         Etobicoke   
100      M9R         Etobicoke   
101      M9V         Etobicoke   
102      M9W         Etobicoke   

                                         Neighbourhood   Latitude  Longitude  
0                                        Rouge,Malvern  43.806686 -79.194353  
1                 Highland Creek,Rouge Hill,Port Union  43.784535 -79.160497  
2                      Guildwood,Morningside,West Hill  43.763573 -79.188711  
3                                               Woburn  43.770992 -79.216917  
4                                            Cedarbrae  43.773136 -79.239476  
5                                  Scarborough Village  43.744734 -79.239476  
6            East Birchmount Park,Ionview,Kennedy Park  43.727929 -79.262029  
7                        Clairlea,Golden Mile,Oakridge  43.711112 -79.284577  
8        Cliffcrest,Cliffside,Scarborough Village West  43.716316 -79.239476  
9                           Birch Cliff,Cliffside West  43.692657 -79.264848  
10   Dorset Park,Scarborough Town Centre,Wexford He...  43.757410 -79.273304  
11                                    Maryvale,Wexford  43.750072 -79.295849  
12                                           Agincourt  43.794200 -79.262029  
13               Clarks Corners,Sullivan,Tam O'Shanter  43.781638 -79.304302  
14   Agincourt North,L'Amoreaux East,Milliken,Steel...  43.815252 -79.284577  
15                                     L'Amoreaux West  43.799525 -79.318389  
16                                         Upper Rouge  43.836125 -79.205636  
17                                   Hillcrest Village  43.803762 -79.363452  
18                          Fairview,Henry Farm,Oriole  43.778517 -79.346556  
19                                     Bayview Village  43.786947 -79.385975  
20                             Silver Hills,York Mills  43.757490 -79.374714  
21                              Newtonbrook,Willowdale  43.789053 -79.408493  
22                                    Willowdale South  43.770120 -79.408493  

importing libraries required for the exploration of the Toronto neighbourhoods

In [53]:
import requests # library to handle requests
import pandas as pd # library for data analsysis
import numpy as np # library to handle data in a vectorized manner
import random # library for random number generation

!conda install -c conda-forge geopy --yes 
from geopy.geocoders import Nominatim # module to convert an address into latitude and longitude values

# libraries for displaying images
from IPython.display import Image 
from IPython.core.display import HTML 
    
# tranforming json file into a pandas dataframe library
from pandas.io.json import json_normalize

!conda install -c conda-forge folium=0.5.0 --yes
import folium # plotting library

print('Folium installed')
print('Libraries imported.')

Fetching package metadata .............
Solving package specifications: .

Package plan for installation in environment C:\Users\abdulmumuni\Anaconda3:

The following NEW packages will be INSTALLED:

    ca-certificates:        2019.6.16-hecc5488_0 conda-forge
    certifi:                2019.6.16-py36_1     conda-forge
    conda-package-handling: 1.4.1-py36_0         conda-forge
    geographiclib:          1.49-py_0            conda-forge
    geopy:                  1.20.0-py_0          conda-forge
    libarchive:             3.3.3-hd1ea8e3_1002  conda-forge
    libiconv:               1.15-hfa6e2cd_1005   conda-forge
    libxml2:                2.9.9-h9ce36c8_2     conda-forge
    lz4-c:                  1.8.1.2-vc14_0       conda-forge
    lzo:                    2.10-hfa6e2cd_1000   conda-forge
    pysocks:                1.7.0-py36_0         conda-forge
    python-libarchive-c:    2.8-py36_1004        conda-forge
    tqdm:                   4.34.0-py_0          conda-forge
    url

ERROR conda.core.link:_execute_actions(337): An error occurred while installing package 'conda-forge::tqdm-4.34.0-py_0'.
CondaError: Cannot link a source that does not exist. C:\Users\abdulmumuni\Anaconda3\Scripts\conda.exe
Attempting to roll back.


CondaError: Cannot link a source that does not exist. C:\Users\abdulmumuni\Anaconda3\Scripts\conda.exe




Fetching package metadata .............
Solving package specifications: .

Package plan for installation in environment C:\Users\abdulmumuni\Anaconda3:

The following NEW packages will be INSTALLED:

    altair:                 3.2.0-py36_0         conda-forge
    blas:                   1.0-mkl                         
    branca:                 0.3.1-py_0           conda-forge
    ca-certificates:        2019.6.16-hecc5488_0 conda-forge
    certifi:                2019.6.16-py36_1     conda-forge
    conda-package-handling: 1.4.1-py36_0         conda-forge
    folium:                 0.5.0-py_0           conda-forge
    libarchive:             3.3.3-hd1ea8e3_1002  conda-forge
    libiconv:               1.15-hfa6e2cd_1005   conda-forge
    libxml2:                2.9.9-h9ce36c8_2     conda-forge
    lz4-c:                  1.8.1.2-vc14_0       conda-forge
    lzo:                    2.10-hfa6e2cd_1000   conda-forge
    pysocks:                1.7.0-py36_0         conda-forge
    pyt

ERROR conda.core.link:_execute_actions(337): An error occurred while installing package 'conda-forge::tqdm-4.34.0-py_0'.
CondaError: Cannot link a source that does not exist. C:\Users\abdulmumuni\Anaconda3\Scripts\conda.exe
Attempting to roll back.

INFO menuinst_win32:__init__(182): Menu: name: 'Anaconda${PY_VER} ${PLATFORM}', prefix: 'C:\Users\abdulmumuni\Anaconda3', env_name: 'None', mode: 'user', used_mode: 'user'

CondaError: Cannot link a source that does not exist. C:\Users\abdulmumuni\Anaconda3\Scripts\conda.exe




In [54]:
CLIENT_ID = 'FGUMY1L22X5NVXOXGK2MVL5V4FYAMTTVZS0NNY5IJRZLD1MI' # your Foursquare ID
CLIENT_SECRET = 'OOKUUUHCFCIOVF2FIZJPWID11XF1JCK2TYBSS2HT55XPNNDL' # your Foursquare Secret
VERSION = '20180604'
LIMIT = 30
print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: FGUMY1L22X5NVXOXGK2MVL5V4FYAMTTVZS0NNY5IJRZLD1MI
CLIENT_SECRET:OOKUUUHCFCIOVF2FIZJPWID11XF1JCK2TYBSS2HT55XPNNDL


checking for latitude and longitude of a neighbourhood, for example Christie Neighbourhood in Downtown Toronto Borough

In [81]:

address = 'Christie, Toronto'

geolocator = Nominatim(user_agent="foursquare_agent")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print(latitude, longitude)

43.6641106 -79.4184051


In [82]:
search_query = 'Chinese'
radius = 500
print(search_query + ' .... OK!')

Chinese .... OK!


In [83]:
url = 'https://api.foursquare.com/v2/venues/search?client_id={}&client_secret={}&ll={},{}&v={}&query={}&radius={}&limit={}'.format(CLIENT_ID, CLIENT_SECRET, latitude, longitude, VERSION, search_query, radius, LIMIT)
url

'https://api.foursquare.com/v2/venues/search?client_id=FGUMY1L22X5NVXOXGK2MVL5V4FYAMTTVZS0NNY5IJRZLD1MI&client_secret=OOKUUUHCFCIOVF2FIZJPWID11XF1JCK2TYBSS2HT55XPNNDL&ll=43.6641106,-79.4184051&v=20180604&query=Chinese&radius=500&limit=30'

Using the GET Request and then examine the results

In [84]:
results = requests.get(url).json()
results

{'meta': {'code': 200, 'requestId': '5d5b44dda30619002ca88ef0'},
 'response': {'venues': [{'categories': [{'icon': {'prefix': 'https://ss3.4sqi.net/img/categories_v2/food/asian_',
       'suffix': '.png'},
      'id': '4bf58dd8d48988d145941735',
      'name': 'Chinese Restaurant',
      'pluralName': 'Chinese Restaurants',
      'primary': True,
      'shortName': 'Chinese'}],
    'hasPerk': False,
    'id': '53bd9f52498e6187cb7c185a',
    'location': {'address': '623 Bloor St W',
     'cc': 'CA',
     'city': 'Toronto',
     'country': 'Canada',
     'distance': 347,
     'formattedAddress': ['623 Bloor St W', 'Toronto ON M6G 1K8', 'Canada'],
     'labeledLatLngs': [{'label': 'display',
       'lat': 43.66451790083401,
       'lng': -79.41412828715735}],
     'lat': 43.66451790083401,
     'lng': -79.41412828715735,
     'postalCode': 'M6G 1K8',
     'state': 'ON'},
    'name': 'Chinese Dumpling House 中华饺子楼',
    'referralId': 'v-1566262493'}]}}

assigning the relevant part of JSON to venues and transform it into a pandas dataframe

In [85]:
# assign relevant part of JSON to venues
venues = results['response']['venues']

# tranform venues into a dataframe
dataframe = json_normalize(venues)
dataframe.head()

categories  hasPerk  \
0  [{'id': '4bf58dd8d48988d145941735', 'name': 'C...    False   

                         id location.address location.cc location.city  \
0  53bd9f52498e6187cb7c185a   623 Bloor St W          CA       Toronto   

  location.country  location.distance  \
0           Canada                347   

                      location.formattedAddress  \
0  [623 Bloor St W, Toronto ON M6G 1K8, Canada]   

                             location.labeledLatLngs  location.lat  \
0  [{'label': 'display', 'lat': 43.66451790083401...     43.664518   

   location.lng location.postalCode location.state  \
0    -79.414128             M6G 1K8             ON   

                           name    referralId  
0  Chinese Dumpling House 中华饺子楼  v-1566262493

Defining information of interest and filtering dataframe

In [86]:
# keep only columns that include venue name, and anything that is associated with location
filtered_columns = ['name', 'categories'] + [col for col in dataframe.columns if col.startswith('location.')] + ['id']
dataframe_filtered = dataframe.loc[:, filtered_columns]

# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

# filter the category for each row
dataframe_filtered['categories'] = dataframe_filtered.apply(get_category_type, axis=1)

# clean column names by keeping only last term
dataframe_filtered.columns = [column.split('.')[-1] for column in dataframe_filtered.columns]

dataframe_filtered

name          categories         address  cc  \
0  Chinese Dumpling House 中华饺子楼  Chinese Restaurant  623 Bloor St W  CA   

      city country  distance                              formattedAddress  \
0  Toronto  Canada       347  [623 Bloor St W, Toronto ON M6G 1K8, Canada]   

                                      labeledLatLngs        lat        lng  \
0  [{'label': 'display', 'lat': 43.66451790083401...  43.664518 -79.414128   

  postalCode state                        id  
0    M6G 1K8    ON  53bd9f52498e6187cb7c185a

Let's visualize the Chinese restaurant that is nearby

In [88]:
dataframe_filtered.name

0    Chinese Dumpling House 中华饺子楼
Name: name, dtype: object

exploring the venue of interest further

In [91]:

venue_id = '53bd9f52498e6187cb7c185a' # ID of Chinese Dumpling House
url = 'https://api.foursquare.com/v2/venues/{}?client_id={}&client_secret={}&v={}'.format(venue_id, CLIENT_ID, CLIENT_SECRET, VERSION)
url

'https://api.foursquare.com/v2/venues/53bd9f52498e6187cb7c185a?client_id=FGUMY1L22X5NVXOXGK2MVL5V4FYAMTTVZS0NNY5IJRZLD1MI&client_secret=OOKUUUHCFCIOVF2FIZJPWID11XF1JCK2TYBSS2HT55XPNNDL&v=20180604'

In [92]:
result = requests.get(url).json()
print(result['response']['venue'].keys())
result['response']['venue']

dict_keys(['id', 'name', 'contact', 'location', 'canonicalUrl', 'categories', 'verified', 'stats', 'price', 'likes', 'dislike', 'ok', 'rating', 'ratingColor', 'ratingSignals', 'allowMenuUrlEdit', 'beenHere', 'specials', 'photos', 'reasons', 'hereNow', 'createdAt', 'tips', 'shortUrl', 'timeZone', 'listed', 'hours', 'pageUpdates', 'inbox', 'attributes', 'bestPhoto', 'colors'])


{'allowMenuUrlEdit': True,
 'attributes': {'groups': [{'count': 1,
    'items': [{'displayName': 'Price', 'displayValue': '$', 'priceTier': 1}],
    'name': 'Price',
    'summary': '$',
    'type': 'price'},
   {'count': 3,
    'items': [{'displayName': 'Reservations', 'displayValue': 'No'}],
    'name': 'Reservations',
    'type': 'reservations'},
   {'count': 1,
    'items': [{'displayName': 'Outdoor Seating', 'displayValue': 'No'}],
    'name': 'Outdoor Seating',
    'type': 'outdoorSeating'},
   {'count': 8,
    'items': [{'displayName': 'Lunch', 'displayValue': 'Lunch'},
     {'displayName': 'Dinner', 'displayValue': 'Dinner'}],
    'name': 'Menus',
    'summary': 'Lunch & Dinner',
    'type': 'serves'},
   {'count': 5,
    'items': [{'displayName': 'Delivery', 'displayValue': 'Delivery'}],
    'name': 'Dining Options',
    'summary': 'Delivery',
    'type': 'diningOptions'}]},
 'beenHere': {'count': 0,
  'lastCheckinExpiredAt': 0,
  'marked': False,
  'unconfirmedCount': 0},
 'be

getting the overall rating of the place

In [93]:
try:
    print(result['response']['venue']['rating'])
except:
    print('This venue has not been rated yet.')

6.9


not a very good rating, now checking the tips

In [94]:
result['response']['venue']['tips']['count']

5

not so many tips, now creating a url and send GET request for the venue tips

In [95]:
## Ecco Tips
limit = 15 # set limit to be greater than or equal to the total number of tips
url = 'https://api.foursquare.com/v2/venues/{}/tips?client_id={}&client_secret={}&v={}&limit={}'.format(venue_id, CLIENT_ID, CLIENT_SECRET, VERSION, limit)

results = requests.get(url).json()
results

{'meta': {'code': 200, 'requestId': '5d5b4d097f797700250a9a07'},
 'response': {'tips': {'count': 5,
   'items': [{'agreeCount': 0,
     'canonicalUrl': 'https://foursquare.com/item/542b73d2498e61689357a7a4',
     'createdAt': 1412133842,
     'disagreeCount': 0,
     'id': '542b73d2498e61689357a7a4',
     'lang': 'en',
     'likes': {'count': 0, 'groups': []},
     'logView': True,
     'text': 'Really good dumplings steamed or fried. Flavourful and home made. We saw a noodle dish that was huge and popular. All at very good prices with friendly staff.',
     'todo': {'count': 0},
     'type': 'user',
     'user': {'firstName': 'Cindy',
      'gender': 'female',
      'id': '24020106',
      'lastName': 'Ross Pedersen',
      'photo': {'prefix': 'https://fastly.4sqi.net/img/user/',
       'suffix': '/G2BZ2EBE1JHDITUH.jpg'}}}]}}}

Getting tips and list of associated features

In [96]:
tips = results['response']['tips']['items']

tip = results['response']['tips']['items'][0]
tip.keys()

dict_keys(['id', 'createdAt', 'text', 'type', 'canonicalUrl', 'lang', 'likes', 'logView', 'agreeCount', 'disagreeCount', 'todo', 'user'])

In [97]:
pd.set_option('display.max_colwidth', -1)

tips_df = json_normalize(tips) # json normalize tips

# columns to keep
filtered_columns = ['text', 'agreeCount', 'disagreeCount', 'id', 'user.firstName', 'user.lastName', 'user.gender', 'user.id']
tips_filtered = tips_df.loc[:, filtered_columns]

# display tips
tips_filtered

text  \
0  Really good dumplings steamed or fried. Flavourful and home made. We saw a noodle dish that was huge and popular. All at very good prices with friendly staff.   

   agreeCount  disagreeCount                        id user.firstName  \
0  0           0              542b73d2498e61689357a7a4  Cindy           

   user.lastName user.gender   user.id  
0  Ross Pedersen  female      24020106

In [100]:
user_id = '24020106' # user ID with most agree counts and complete profile

url = 'https://api.foursquare.com/v2/users/{}?client_id={}&client_secret={}&v={}'.format(user_id, CLIENT_ID, CLIENT_SECRET, VERSION) # define URL

# send GET request
results = requests.get(url).json()
user_data = results['response']['user']

# display features associated with user
user_data.keys()

dict_keys(['id', 'firstName', 'lastName', 'gender', 'canonicalUrl', 'photo', 'friends', 'tips', 'homeCity', 'bio', 'contact', 'photos', 'type', 'mayorships', 'checkins', 'lists', 'lenses'])

In [101]:
print('First Name: ' + user_data['firstName'])
print('Last Name: ' + user_data['lastName'])
print('Home City: ' + user_data['homeCity'])

First Name: Cindy
Last Name: Ross Pedersen
Home City: Toronto, Ontario


Getting user´s information 

In [102]:
user_data['tips']

{'count': 48}

In [103]:
# define tips URL
url = 'https://api.foursquare.com/v2/users/{}/tips?client_id={}&client_secret={}&v={}&limit={}'.format(user_id, CLIENT_ID, CLIENT_SECRET, VERSION, limit)

# send GET request and get user's tips
results = requests.get(url).json()
tips = results['response']['tips']['items']

# format column width
pd.set_option('display.max_colwidth', -1)

tips_df = json_normalize(tips)

# filter columns
filtered_columns = ['text', 'agreeCount', 'disagreeCount', 'id']
tips_filtered = tips_df.loc[:, filtered_columns]

# display user's tips
tips_filtered

text  \
0   Yoga outdoors in the summer!                                                                                                                                                                 
1   Food is good.  Service is good.  Prices reasonable. Private spaces if you need them. Everything you expect from a primarily northern  indian style restaurant in a quiet location.           
2   Great place for a meet up any time of day. Morning coffee and treat, mid day drink and snack, after work cocktail and shared plate.                                                          
3   Servings are big so don't be hesitant about asking for the small size.                                                                                                                       
4   Great haircuts, super cheap. Why spend $60-$100 when you can spend $10.                                                                                                                      
5   Cool and cosy spot. Massive oven. Great home made bread. Very good pasta. Special veggie menu. You could spend a couple of weeks working through the fun restaurants in this neighborhood.   
6   I concur...the cauliflower is amazing.                                                                                                                                                       
7   Great place for groups to gather for dinner (friends or biz). Lots of interesting tables in the kitchen and bar. Food is interesting and tasting menu is reasonable. Hidden treasure.        
8   Really good dumplings steamed or fried. Flavourful and home made. We saw a noodle dish that was huge and popular. All at very good prices with friendly staff.                               
9   Only homogenized milk fot coffee. Not customer centric!                                                                                                                                      
10  Thank you Rotman! Finally a great espresso on campus (and some good lookin' food too).                                                                                                       
11  Fantastic sweets and savories and a great espresso to go with it. Limited seating.                                                                                                           
12  People come for the drinks but the food is better than you might think. And all the desserts made in house, quite yummy. I had great service today, thanks to hostess Ginger.                
13  OMG great jerk chicken done on the bbq outside the store. I challenge you to pass by without getting some.                                                                                   
14  Always awesome. Great food and great desserts. Lovely ambiance. Just a little iffy on the service today.                                                                                     

    agreeCount  disagreeCount                        id  
0   0           0              59613a4cea1e44469f470912  
1   0           0              57c32f33498ef82d29e00647  
2   0           0              579b5a2bcd10ba9158cb778c  
3   0           0              568d8561498ed4aaf0111ff8  
4   0           0              568d853f498e9a34fe0ebee0  
5   0           0              54e9f750498eaa43b110c1e4  
6   0           0              54e9f6bc498edb052b6a5feb  
7   0           0              54e9f69b498eaa43b110b7d2  
8   0           0              542b73d2498e61689357a7a4  
9   0           0              542b0da4498e17b325abb93b  
10  0           0              52961ce411d208de272d3fe8  
11  0           0              525bf56411d207c6281fb8a3  
12  0           0              520a82e811d257e4eeb2f143  
13  0           0              51eafa6f498e70eec39c76b6  
14  0           0              51866b1c498e42f65967cc28

In [107]:
tip_id = '542b73d2498e61689357a7a4' # tip id

# define URL
url = 'http://api.foursquare.com/v2/tips/{}?client_id={}&client_secret={}&v={}'.format(tip_id, CLIENT_ID, CLIENT_SECRET, VERSION)

# send GET Request and examine results
result = requests.get(url).json()
print(result['response']['tip']['venue']['name'])
print(result['response']['tip']['venue']['location'])

Chinese Dumpling House 中华饺子楼
{'address': '623 Bloor St W', 'lat': 43.66451790083401, 'lng': -79.41412828715735, 'labeledLatLngs': [{'label': 'display', 'lat': 43.66451790083401, 'lng': -79.41412828715735}], 'postalCode': 'M6G 1K8', 'cc': 'CA', 'city': 'Toronto', 'state': 'ON', 'country': 'Canada', 'formattedAddress': ['623 Bloor St W', 'Toronto ON M6G 1K8', 'Canada']}


In [108]:
user_data

{'bio': '',
 'canonicalUrl': 'https://foursquare.com/cindyrp',
 'checkins': {'count': 517, 'items': []},
 'contact': {},
 'firstName': 'Cindy',
 'friends': {'count': 26,
  'groups': [{'count': 26,
    'items': [{'bio': '',
      'contact': {},
      'firstName': 'Candy',
      'gender': 'female',
      'homeCity': 'Dijon, A1',
      'id': '80066560',
      'lastName': 'Pauline',
      'lists': {'groups': [{'count': 2, 'items': [], 'type': 'created'}]},
      'photo': {'prefix': 'https://fastly.4sqi.net/img/user/',
       'suffix': '/80066560-UL2H2BOK2TSHMESP.jpg'},
      'tips': {'count': 3}},
     {'bio': '',
      'contact': {},
      'firstName': 'Kay',
      'gender': 'female',
      'homeCity': 'Toronto, ON',
      'id': '91089389',
      'lastName': 'Bian',
      'lists': {'groups': [{'count': 2, 'items': [], 'type': 'created'}]},
      'photo': {'prefix': 'https://fastly.4sqi.net/img/user/',
       'suffix': '/91089389-UCWEH1KXVL55FHDX.jpg'},
      'tips': {'count': 0}},
     {'

In [106]:
# 1. grab prefix of photo
# 2. grab suffix of photo
# 3. concatenate them using the image size  
Image(url='https://igx.4sqi.net/img/user/300x300/80066560-UL2H2BOK2TSHMESP.jpg')

Exploring shops and other places of interest around the "Chinese Dumpling House"

In [109]:
latitude = 43.664518
longitude = -79.414128

In [110]:
url = 'https://api.foursquare.com/v2/venues/explore?client_id={}&client_secret={}&ll={},{}&v={}&radius={}&limit={}'.format(CLIENT_ID, CLIENT_SECRET, latitude, longitude, VERSION, radius, LIMIT)
url

'https://api.foursquare.com/v2/venues/explore?client_id=FGUMY1L22X5NVXOXGK2MVL5V4FYAMTTVZS0NNY5IJRZLD1MI&client_secret=OOKUUUHCFCIOVF2FIZJPWID11XF1JCK2TYBSS2HT55XPNNDL&ll=43.664518,-79.414128&v=20180604&radius=500&limit=30'

In [111]:
results = requests.get(url).json()
'There are {} around Chinese Dumpling House.'.format(len(results['response']['groups'][0]['items']))

'There are 30 around Chinese Dumpling House.'

In [112]:
items = results['response']['groups'][0]['items']
items[0]

{'reasons': {'count': 0,
  'items': [{'reasonName': 'globalInteractionReason',
    'summary': 'This spot is popular',
    'type': 'general'}]},
 'referralId': 'e-0-4c30e19b213c2d7f9c24335d-0',
 'venue': {'categories': [{'icon': {'prefix': 'https://ss3.4sqi.net/img/categories_v2/shops/spa_',
     'suffix': '.png'},
    'id': '4bf58dd8d48988d1ed941735',
    'name': 'Spa',
    'pluralName': 'Spas',
    'primary': True,
    'shortName': 'Spa'}],
  'id': '4c30e19b213c2d7f9c24335d',
  'location': {'address': '588 Bloor St. W',
   'cc': 'CA',
   'city': 'Toronto',
   'country': 'Canada',
   'crossStreet': 'Bathurst',
   'distance': 118,
   'formattedAddress': ['588 Bloor St. W (Bathurst)',
    'Toronto ON M6G 1K4',
    'Canada'],
   'labeledLatLngs': [{'label': 'display',
     'lat': 43.66483011797359,
     'lng': -79.41272601896088}],
   'lat': 43.66483011797359,
   'lng': -79.41272601896088,
   'postalCode': 'M6G 1K4',
   'state': 'ON'},
  'name': 'Aroma Wellness Clinic and Spa',
  'photos'

In [113]:
dataframe = json_normalize(items) # flatten JSON

# filter columns
filtered_columns = ['venue.name', 'venue.categories'] + [col for col in dataframe.columns if col.startswith('venue.location.')] + ['venue.id']
dataframe_filtered = dataframe.loc[:, filtered_columns]

# filter the category for each row
dataframe_filtered['venue.categories'] = dataframe_filtered.apply(get_category_type, axis=1)

# clean columns
dataframe_filtered.columns = [col.split('.')[-1] for col in dataframe_filtered.columns]

dataframe_filtered.head(10)

name         categories  \
0  Aroma Wellness Clinic and Spa         Spa                 
1  Rustle And Still                      Sandwich Place      
2  Hodo Kwaja 호도과자                       Dessert Shop        
3  Snakes & Lattes                       Café                
4  Herbs & Nutrition Qi Natural Food     Health Food Store   
5  Sunrise House Korean Restaurant 해뜨는집  Korean Restaurant   
6  The Good Neighbour                    Coffee Shop         
7  Buk Chang Dong Soon Tofu              Korean Restaurant   
8  Insomnia Restaurant and Lounge        Restaurant          
9  Yummy Korean Restaurant               Korean Restaurant   

                 address  cc     city country  \
0  588 Bloor St. W        CA  Toronto  Canada   
1  605 Bloor Street West  CA  Toronto  Canada   
2  656 Bloor St W.        CA  Toronto  Canada   
3  600 Bloor St. W        CA  Toronto  Canada   
4  572 Bloor St W         CA  Toronto  Canada   
5  661 Bloor Street West  CA  Toronto  Canada   
6  678 Bloor St W         CA  Toronto  Canada   
7  691 Bloor St. W.       CA  Toronto  Canada   
8  563 Bloor St West      CA  Toronto  Canada   
9  620 Bloor Street West  CA  Toronto  Canada   

                            crossStreet  distance  \
0  Bathurst                              118        
1  NaN                                   85         
2  NaN                                   120        
3  btwn Markham St. & Palmerston Ave.    84         
4  at Bathurst                           187        
5  Manning                               135        
6  Manning                               210        
7  at Clinton St.                        250        
8  Bathurst St                           265        
9  btwn.  Elucid Ave. & Palmerston Ave.  18         

                                                                             formattedAddress  \
0  [588 Bloor St. W (Bathurst), Toronto ON M6G 1K4, Canada]                                     
1  [605 Bloor Street West, Toronto ON M6G 1K4, Canada]                                          
2  [656 Bloor St W., Toronto ON, Canada]                                                        
3  [600 Bloor St. W (btwn Markham St. & Palmerston Ave.), Toronto ON M6G 1K4, Canada]           
4  [572 Bloor St W (at Bathurst), Toronto ON M6G 1K1, Canada]                                   
5  [661 Bloor Street West (Manning), Toronto ON, Canada]                                        
6  [678 Bloor St W (Manning), Toronto ON, Canada]                                               
7  [691 Bloor St. W. (at Clinton St.), Toronto ON M6G 1L3, Canada]                              
8  [563 Bloor St West (Bathurst St), Toronto ON M5S 1Y6, Canada]                                
9  [620 Bloor Street West (btwn.  Elucid Ave. & Palmerston Ave.), Toronto ON M6G 1K9, Canada]   

                                                                 labeledLatLngs  \
0  [{'label': 'display', 'lat': 43.66483011797359, 'lng': -79.41272601896088}]    
1  [{'label': 'display', 'lat': 43.66473385467293, 'lng': -79.41310970667249}]    
2  [{'label': 'display', 'lat': 43.6642403221846, 'lng': -79.41557903826563}]     
3  [{'label': 'display', 'lat': 43.66478921378291, 'lng': -79.41315049188894}]    
4  [{'label': 'display', 'lat': 43.665013428045555, 'lng': -79.41189818065214}]   
5  [{'label': 'display', 'lat': 43.66418287914217, 'lng': -79.41574611551155}]    
6  [{'label': 'display', 'lat': 43.66398338946789, 'lng': -79.41664049314092}]    
7  [{'label': 'display', 'lat': 43.663841985166, 'lng': -79.41709282188468}]      
8  [{'label': 'display', 'lat': 43.665179618013326, 'lng': -79.41096552569321}]   
9  [{'label': 'display', 'lat': 43.664430691694356, 'lng': -79.41393731132068}]   

         lat        lng neighborhood postalCode state  \
0  43.664830 -79.412726  NaN          M6G 1K4    ON     
1  43.664734 -79.413110  Koreatown    M6G 1K4    ON     
2  43.664240 -79.415579  NaN          NaN        ON     
3  43.664789 -79.4